src/blog/understanding-ecmascript-part-3.md

```diff
@@ -0,0 +1,320 @@
+
+The subscripts are a shorthand for expressing a set of productions, for a set of left-hand side symbols, all at once. The left-hand side symbol has two parameters, so the "real" left-hand side symbols we're defining are `VariableStatement`, `VariableStatement_Yield`, `VariableStatement_Await` and `VariableStatement_Yield_Await`.
+
+Note that here the plain `VariableStatement` means "`VariableStatement` without `_Await` and without `_Yield`" and shoud not be confused with <code>VariableStatement<sub>[Yield, Await]</sub></code>.
+
+On the right-hand side of the production, we see the shorthand `+In`, meaning "use the version with `_In`", and `?Await`, meaning "use the version with `_Await` iff the left-hand side symbol has `_Await` (similarly with `?Yield`).
+
+(The third shorthand, `~Foo`, meaning "use the version without `_Foo`", is not used in this production.)
+
+With this information, we can expand the productions like this:
+
+> `VariableStatement` :
+> `var VariableDeclarationList_In;`
+>
+> `VariableStatement_Yield` :
+> `var VariableDeclarationList_In_Yield;`
+>
+> `VariableStatement_Await` :
+> `var VariableDeclarationList_In_Await;`
+>
+> `VariableStatement_Yield_Await` :
+> `var VariableDeclarationList_In_Yield_Await;`
+
+Ultimately, we'll need to find out two things: 1) Where is it decided whether we're in the case with `_Await` or without `_Await`? 2) Where does it make a difference &mdash; where do the productions for `Something_Await` and `Something` (without `_Await`) diverge?
+
+### `_Await` or no `_Await`?
+
+Let's tackle question 1 first. It's somewhat easy to guess that non-async functions and async functions differ in whether we pick the parameter `_Await` for the function body or not. Reading the productions for async function declarations, we find this:
+
+> [`AsyncFunctionBody :`](https://tc39.es/ecma262/#prod-AsyncFunctionBody)
+> <code>FunctionBody<sub>[~Yield, +Await]</sub></code>
+
+Note that `AsyncFunctionBody` has no parameters &mdash; they get added to the `FunctionBody` on the right-hand side.
+
+If we expand this production, we get:
+
+> `AsyncFunctionBody :`
+> `FunctionBody_Await`
+
+Since `FunctionBody_Await` is used for async functions. It means a function body where `await` is treated as a keyword.
+
+On the other hand, if we're inside a non-async function, the relevant production is:
+
+> [<code>FunctionDeclaration<sub>[Yield, Await, Default]</sub>](https://tc39.es/ecma262/#prod-FunctionDeclaration) :</code>
+> <code>function BindingIdentifier<sub>[?Yield, ?Await]</sub> ( FormalParameters<sub>[~Yield, ~Await]</sub> ) { FunctionBody<sub>[~Yield, ~Await]</sub> }</code>
+
+(`FunctionDeclaration` has another production, but it's not relevant for our code example.)
+
+To avoid combinatorial expansion, let's ignore the `Default` parameter which is not used in this particular production.
+
+The expanded form of the production is:
```

Comment:
Either there is a typo here, or I don't get it. I don't see a `FunctionBody_Await` here.

Reply:
The next paragraph just after the productions explains it:

The important thing in this production is that FunctionBody is parameterized with [~Yield, ~Await], meaning that we always take the version without _Yield and without _Await, no matter what the parameters of FunctionDeclaration were.

Reply:
Rephrased:

In this production we always get FunctionBody (without _Yield and without _Await), since the FunctionBody in the non-expanded production is parameterized with [~Yield, ~Await].